compiler/rustc_middle/src/mir/interpret/allocation/provenance_map.rs

```diff
@@ -212,23 +212,36 @@ impl<Prov: Provenance> ProvenanceMap<Prov> {
         Ok(())
     }
 
-    /// Overwrites all provenance in the allocation with wildcard provenance.
+    /// Overwrites all provenance in the given range with wildcard provenance.
+    /// Pointers partially overwritten will have their provenances
     ///
     /// Provided for usage in Miri and panics otherwise.
-    pub fn write_wildcards(&mut self, alloc_size: usize) {
+    pub fn write_wildcards(&mut self, cx: &impl HasDataLayout, range: AllocRange) {
         assert!(
             Prov::OFFSET_IS_ADDR,
             "writing wildcard provenance is not supported when `OFFSET_IS_ADDR` is false"
         );
         let wildcard = Prov::WILDCARD.unwrap();
 
-        // Remove all pointer provenances, then write wildcards into the whole byte range.
-        self.ptrs.clear();
-        let last = Size::from_bytes(alloc_size);
+        // Get pointer provenances that overlap with the range, then remove them.
+        let ptr_range = Self::adjusted_range_ptrs(range, cx);
+        let removed = ptr_range.filter_map(|i| self.ptrs.remove(&i).map(|p| (i, p)));
```

Comment:
FWIW, `remove_range` would probably be more efficient. That doesn't return all the removed elements, but you anyway don't want *all* of them, just the first and last. So I think this is better:
```rust
// Preserve partially overwritten pointers.
{
  let ptrs = self.ptrs.range(ptr_range);
  if let Some((offset, prov)) = ptrs.first() {
    for byte_ofs in offset..range.start { bytes.insert(byte_ofs, prov); }
  }
  if let Some((offset, prov)) = ptrs.last() {
    for byte_ofs in range.end()..offset + cx.data_layout().pointer_size() { bytes.insert(byte_ofs, prov); }
  }
}
// Now this can all be removed.
self.ptrs.remove_range(ptr_range);
// Overwrite bytewise provenance.
...
```